core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -202,8 +205,11 @@ private[spark] class TaskSchedulerImpl(
     val tasks = taskSet.tasks
     logInfo("Adding task set " + taskSet.id + " with " + tasks.length + " tasks")
     this.synchronized {
-      val manager = createTaskSetManager(taskSet, maxTaskFailures)
       val stage = taskSet.stageId
+      // only create a BitSet once for a certain stage since we only remove
+      // that stage when an active TaskSetManager succeed.
+      stageIdToFinishedPartitions.getOrElseUpdate(stage, new BitSet)
```

Comment:
nit: can we move it to `createTaskSetManager`, and keep `submitTasks` untouched?

Reply:
updated.